File: packages/typia/src/programmers/internal/check_number.ts
import ts from "typescript";

import { ExpressionFactory } from "../../factories/ExpressionFactory";

import { MetadataAtomic } from "../../schemas/metadata/MetadataAtomic";

import { ITypiaContext } from "../../transformers/ITypiaContext";

import { ICheckEntry } from "../helpers/ICheckEntry";
import { OptionPredicator } from "../helpers/OptionPredicator";

/** @internal */
export const check_number = (props: {
  numeric: boolean;
  context: ITypiaContext;
  atomic: MetadataAtomic;
  input: ts.Expression;
}): ICheckEntry => {
  const base: ts.BinaryExpression = ts.factory.createStrictEquality(
    ts.factory.createStringLiteral("number"),
    ts.factory.createTypeOfExpression(props.input),
  );
  const addition: ts.Expression | null =
    props.numeric === true
      ? OptionPredicator.finite(props.context.options)
        ? ts.factory.createCallExpression(
            ts.factory.createIdentifier("Number.isFinite"),
            undefined,
            [props.input],
          )
        : OptionPredicator.numeric(props.context.options)
          ? ts.factory.createLogicalNot(
              ts.factory.createCallExpression(
                ts.factory.createIdentifier("Number.isNaN"),
                undefined,
                [props.input],
              ),
            )
          : null
      : null;
  const conditions: ICheckEntry.ICondition[][] = check_numeric_type_tags({
    context: props.context,
    atomic: props.atomic,
    input: props.input,
    addition,
  });
  return {
    expected: props.atomic.getName(),
    expression:
      addition !== null && conditions.length === 0
        ? ts.factory.createLogicalAnd(base, addition)
        : base,
    conditions,
  };
};

/** @internal */
const check_numeric_type_tags = (props: {
  context: ITypiaContext;
  atomic: MetadataAtomic;
  addition: ts.Expression | null;
  input: ts.Expression;
}): ICheckEntry.ICondition[][] =>
  props.atomic.tags
    .map((row) => row.filter((tag) => !!tag.validate))
    .filter((row) => !!row.length)
    .map((row) => [
      ...(props.addition === null
        ? []
        : row.some(
              (tag) =>
                tag.kind === "type" &&
                (tag.value === "int32" ||
                  tag.value === "uint32" ||
                  tag.value === "int64" ||
                  tag.value === "uint64" ||
                  tag.value === "float"),
            ) ||
            row.some(
              (tag) =>
                tag.kind === "multipleOf" && typeof tag.value === "number",
            ) ||
            (row.some(
              (tag) =>
                (tag.kind === "minimum" || tag.kind === "exclusiveMinimum") &&
                typeof tag.value === "number",
            ) &&
              row.some(
                (tag) =>
                  (tag.kind === "maximum" || tag.kind === "exclusiveMaximum") &&
                  typeof tag.value === "number",
              ))
          ? []
          : [
              {
                expected: "number",
                expression: props.addition!,
              },
            ]),
      ...row.map((tag) => ({
        expected: `number & ${tag.name}`,
        expression: ExpressionFactory.transpile({
          transformer: props.context.transformer,
          importer: props.context.importer,
          script: tag.validate!,
        })(props.input),
      })),
    ]);

File: packages/typia/native/core/programmers/iterate/check_number.go
package iterate

import (
	"fmt"

	shimast "github.com/microsoft/typescript-go/shim/ast"
	nativecontext "github.com/samchon/typia/packages/typia/native/core/context"
	nativefactories "github.com/samchon/typia/packages/typia/native/core/factories"
	nativehelpers "github.com/samchon/typia/packages/typia/native/core/programmers/helpers"
	nativemetadata "github.com/samchon/typia/packages/typia/native/core/schemas/metadata"
)

type Check_numberProps struct {
	Numeric bool
	Context nativecontext.ITypiaContext
	Atomic  *nativemetadata.MetadataAtomic
	Input   *shimast.Expression
}

func Check_number(props Check_numberProps) nativehelpers.ICheckEntry {
	base := check_number_factory.NewBinaryExpression(
		nil,
		check_number_factory.NewStringLiteral("number", shimast.TokenFlagsNone),
		nil,
		check_number_factory.NewToken(shimast.KindEqualsEqualsEqualsToken),
		check_number_factory.NewTypeOfExpression(props.Input),
	)
	var addition *shimast.Node
	if props.Numeric {
		if nativehelpers.OptionPredicator.Finite(props.Context.Options) {
			addition = check_number_factory.NewCallExpression(
				check_number_factory.NewIdentifier("Number.isFinite"),
				nil,
				nil,
				check_number_factory.NewNodeList([]*shimast.Node{props.Input}),
				shimast.NodeFlagsNone,
			)
		} else if nativehelpers.OptionPredicator.Numeric(props.Context.Options) {
			addition = check_number_factory.NewPrefixUnaryExpression(
				shimast.KindExclamationToken,
				check_number_factory.NewCallExpression(
					check_number_factory.NewIdentifier("Number.isNaN"),
					nil,
					nil,
					check_number_factory.NewNodeList([]*shimast.Node{props.Input}),
					shimast.NodeFlagsNone,
				),
			)
		}
	}
	conditions := check_numeric_type_tags(check_numeric_type_tagsProps{
		Context:  props.Context,
		Atomic:   props.Atomic,
		Input:    props.Input,
		Addition: addition,
	})
	expression := base
	if addition != nil && len(conditions) == 0 {
		expression = check_number_factory.NewBinaryExpression(
			nil,
			base,
			nil,
			check_number_factory.NewToken(shimast.KindAmpersandAmpersandToken),
			addition,
		)
	}
	return nativehelpers.ICheckEntry{
		Expected:   props.Atomic.GetName(),
		Expression: expression,
		Conditions: conditions,
	}
}

type check_numeric_type_tagsProps struct {
	Context  nativecontext.ITypiaContext
	Atomic   *nativemetadata.MetadataAtomic
	Addition *shimast.Expression
	Input    *shimast.Expression
}

func check_numeric_type_tags(props check_numeric_type_tagsProps) [][]nativehelpers.ICheckEntry_ICondition {
	output := [][]nativehelpers.ICheckEntry_ICondition{}
	for _, row := range props.Atomic.Tags {
		tags := check_number_filter_validate(row)
		if len(tags) == 0 {
			continue
		}
		conditions := []nativehelpers.ICheckEntry_ICondition{}
		if props.Addition != nil && !check_numeric_type_tags_covers_addition(tags) {
			conditions = append(conditions, nativehelpers.ICheckEntry_ICondition{
				Expected:   "number",
				Expression: props.Addition,
			})
		}
		for _, tag := range tags {
			conditions = append(conditions, nativehelpers.ICheckEntry_ICondition{
				Expected:   fmt.Sprintf("number & %s", tag.Name),
				Expression: check_number_transpile(props.Context, tag.Validate)(props.Input),
			})
		}
		output = append(output, conditions)
	}
	return output
}

func check_numeric_type_tags_covers_addition(row []nativemetadata.IMetadataTypeTag) bool {
	if check_number_some(row, func(tag nativemetadata.IMetadataTypeTag) bool {
		return tag.Kind == "type" &&
			(tag.Value == "int32" ||
				tag.Value == "uint32" ||
				tag.Value == "int64" ||
				tag.Value == "uint64" ||
				tag.Value == "float")
	}) {
		return true
	}
	if check_number_some(row, func(tag nativemetadata.IMetadataTypeTag) bool {
		return tag.Kind == "multipleOf" && check_number_is_number(tag.Value)
	}) {
		return true
	}
	return check_number_some(row, func(tag nativemetadata.IMetadataTypeTag) bool {
		return (tag.Kind == "minimum" || tag.Kind == "exclusiveMinimum") && check_number_is_number(tag.Value)
	}) && check_number_some(row, func(tag nativemetadata.IMetadataTypeTag) bool {
		return (tag.Kind == "maximum" || tag.Kind == "exclusiveMaximum") && check_number_is_number(tag.Value)
	})
}

func check_number_filter_validate(row []nativemetadata.IMetadataTypeTag) []nativemetadata.IMetadataTypeTag {
	tags := []nativemetadata.IMetadataTypeTag{}
	for _, tag := range row {
		if tag.Validate != "" {
			tags = append(tags, tag)
		}
	}
	return tags
}

func check_number_some(row []nativemetadata.IMetadataTypeTag, pred func(tag nativemetadata.IMetadataTypeTag) bool) bool {
	for _, tag := range row {
		if pred(tag) {
			return true
		}
	}
	return false
}

func check_number_is_number(value any) bool {
	switch value.(type) {
	case int, int8, int16, int32, int64, uint, uint8, uint16, uint32, uint64, float32, float64:
		return true
	default:
		return false
	}
}

func check_number_transpile(context nativecontext.ITypiaContext, script string) func(input *shimast.Expression) *shimast.Node {
	var importer nativefactories.ExpressionFactory_Importer
	if v, ok := context.Importer.(nativefactories.ExpressionFactory_Importer); ok {
		importer = v
	}
	return nativefactories.ExpressionFactory.Transpile(nativefactories.ExpressionFactory_TranspileProps{
		Transformer: context.Transformer,
		Importer:    importer,
		Script:      script,
	})
}

var check_number_factory = shimast.NewNodeFactory(shimast.NodeFactoryHooks{})
